### PR TITLE
Test async pipeline creation before/after device destroy

### LINKED ISSUE #2265: val: state,device_lost,destroy:{createRenderPipelineAsync, createComputePipelineAsync}
Expand these validation tests to test performing async pipeline compilation after the device has been lost.

### PR DESCRIPTION
**Pull request by austinEng**
Fixes #2265

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/api/validation/state/device_lost/destroy.spec.ts [hunk 1]
```diff
@@ -454,6 +454,156 @@ Tests creating render pipeline on destroyed device.
...
+
+    // Destroy the device, and expect it to be lost.
+    t.expectDeviceLost('destroyed');
+    t.device.destroy();
+    if (awaitLost) {
+      const lostInfo = await t.device.lost;
+      t.expect(lostInfo.reason === 'destroyed');
+    }
+
+    // After device destroy, expect a GPUPipelineError with "validation" reason.
```
Comment (lokokung) [inline src/webgpu/api/validation/state/device_lost/destroy.spec.ts:485]: @kainino0x Want to confirm with you whether this is the intended behavior? W.r.t the other APIs, I think that this should actually just return a pipeline? If so, maybe the `during` variants can be combined.

Comment (kainino0x) [reply src/webgpu/api/validation/state/device_lost/destroy.spec.ts:485]: The intent is that it returns a pipeline, not rejects. Although reading the spec closely I think the spec is actually wrong and says it'll reject. Reopened https://github.com/gpuweb/gpuweb/issues/1629#issuecomment-1599988786

Comment (austinEng) [reply src/webgpu/api/validation/state/device_lost/destroy.spec.ts]: I've updated these tests to now expect success for all the destroyed cases

File: src/webgpu/api/validation/state/device_lost/destroy.spec.ts [hunk 2]
```diff
@@ -454,6 +454,142 @@ Tests creating render pipeline on destroyed device.
     }, awaitLost);
   });
 
+g.test('createComputePipelineAsync')
```
Comment (lokokung) [inline src/webgpu/api/validation/state/device_lost/destroy.spec.ts:457]: WDYT about merging the variants with a `awaitLost` param?

Comment (austinEng) [reply src/webgpu/api/validation/state/device_lost/destroy.spec.ts:451]: Merged.

I also ended up turning `awaitLost` into a param not a subcase param. I realized that as a subcase param, it was reusing the device between subcases! which meant that the lost device was already lost before .destroy()'ed in some subcases

Comment (austinEng) [reply src/webgpu/api/validation/state/device_lost/destroy.spec.ts:451]: That is - overall, anything that destroys the device shouldn't use subcases, I think.

**Comment by austinEng** [comment]
@kainino0x is validation still supposed to run on lost devices?

I should probably add test cases that do:
1. destroy the device
2. async pipeline creation with *invalid* WGSL source
3. expect validation error? or "success" ?

File: src/webgpu/api/validation/state/device_lost/destroy.spec.ts (hunk 2 above)
Comment (lokokung) [reply src/webgpu/api/validation/state/device_lost/destroy.spec.ts:451]: Oh great catch!

**Review by lokokung — APPROVED** [review]
Thanks Austin!

**Comment by kainino0x** [comment]
> @kainino0x is validation still supposed to run on lost devices?

No, it should never be observable whether validation is running or not on a lost device. https://gpuweb.github.io/gpuweb/#errors-and-debugging

**Comment by kainino0x** [comment]
and https://github.com/gpuweb/gpuweb/issues/1629#issuecomment-1495166485

**Comment by austinEng** [comment]
> > @kainino0x is validation still supposed to run on lost devices?
> 
> No, it should never be observable whether validation is running or not on a lost device. https://gpuweb.github.io/gpuweb/#errors-and-debugging

Ok, @kainino0x PTAL again

I'm not sure how we should handle invalid creation while losing the device though.
I believe this coudl be a race since the validation could happen before or after the device was destroyed.
If it happened before, you see the error. If it happened after, you don't.
Right now, I have the test written assuming that validation occurs on the same timeline as device.destroy() (not on some thread), so if you do something invalid before device.destroy(), you get the error.

**Comment by kainino0x** [comment]
I think it's a race, yeah. But maybe we can rely on the ordering of the `lost` promise and the createPipelineAsync promise matching the result you see. explained more here: https://github.com/gpuweb/gpuweb/issues/1629#issuecomment-1603468868
and possible test code:

```
let isLost = false;
const lostP = device.lost.then(() => { isLost = true; });

let createSucceeded;
const createP = createPipelineAsync().then(() => {
  assert(isLost);
}, () => {
  assert(!isLost);
});

device.destroy();
await Promise.all([lostP, createP]);
```

wdyt?
Unfortunately probably only one of these paths will actually be hit by chrome.

**Comment by kainino0x** [comment]
oh and i haven't actually reviewed yet so apologies if this is close to what you have. Will review momentarily

**Comment by austinEng** [comment]
@kainino0x updated to handle the validation ordering with the device lost promise

**Review by kainino0x — APPROVED** [review]

**Comment by kainino0x** [comment]
assuming you've gotten the test to pass with the chrome patch, lgtm

**Pull request merged**